x-pack/filebeat/input/awss3/sqs_input.go

```diff
@@ -226,19 +270,23 @@ func (w *sqsWorker) processMessage(ctx context.Context, msg types.Message) {
 		w.client.Publish(e)
 		publishCount++
 	})
+	w.pending.Add(int64(publishCount))
 
 	if publishCount == 0 {
 		// No events made it through (probably an error state), wrap up immediately
 		result.Done()
 	} else {
 		// Add this result's Done callback to the pending ACKs list
-		w.ackHandler.Add(publishCount, result.Done)
+		w.ackHandler.Add(publishCount, func() {
+			result.Done()
+			w.pending.Add(-1)
```

Comment:
Wouldn't this callback of the `ackHandler` run once when the pending queue is empty? (referring to this code segment [1]) . This means the atomic `pending` wouldn't be empty. 

[1] - https://github.com/elastic/beats/blob/v8.17.3/x-pack/filebeat/input/awss3/acks.go#L96-L98

Reply:
Yeah. `(*awsACKHandler).Add` probably should have documentation explaining how to hold it right. The line here needs to either be changed to be `-int64(publishCount)`, or more likely above should just be `1`.

Reply:
You could use an atomic boolean (`atomic.Bool`) instead

Reply:
I'd rather use a counter.